### PR TITLE
net: wifi: shell: add wps support

### PR DESCRIPTION
**Pull request by Rex-Chen-NXP**
Add wps pin and wps pbc L2 layer cmd support.

### CODE REVIEW

**Comment by Rex-Chen-NXP** [comment]
@jukkar  @krish2718 
Could you help review this pr?
By the way, no white space at start of the line 1634 of subsys/net/l2/wifi/wifi_shell.c, what should I change to fix this format warning?

File: subsys/net/l2/wifi/wifi_shell.c [hunk 1]
```diff
@@ -1594,6 +1594,49 @@ static int cmd_wifi_btm_query(const struct shell *sh, size_t argc, char *argv[])
...
+		return -ENOEXEC;
+	}
+
+	if (net_mgmt(NET_REQUEST_WIFI_WPS_PIN, iface, &params, sizeof(params))) {
+		PR_WARNING("Start wps pbc connection failed\n");
+		return -ENOEXEC;
+	}
+
+	if (params.oper == WIFI_WPS_PIN_GET) {
+		PR("WPS PIN is: %s \n", params.get_pin);
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_shell.c:1634]:
```suggestion
		PR("WPS PIN is: %s\n", params.get_pin);
```

File: subsys/net/l2/wifi/wifi_shell.c [hunk 2]
```diff
@@ -2512,10 +2555,10 @@ SHELL_STATIC_SUBCMD_SET_CREATE(
...
 );
 
-SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
-	SHELL_CMD_ARG(version, NULL, "Print Wi-Fi Driver and Firmware versions\n",
-		  cmd_wifi_version,
-		  1, 0),
+SHELL_STATIC_SUBCMD_SET_CREATE(
+	wifi_commands,
+	SHELL_CMD_ARG(version, NULL, "Print Wi-Fi Driver and Firmware versions\n", cmd_wifi_version,
+		      1, 0),
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_shell.c:2561]: whitespace change, please remove or move to a separate commit.

File: subsys/net/l2/wifi/wifi_shell.c [hunk 3]
```diff
@@ -2625,6 +2668,13 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
 		cmd_wifi_btm_query,
 		2, 0),
 #endif
+	SHELL_CMD_ARG(wps_pbc, NULL,
+		"Start wps pbc connection.\n",
+		cmd_wifi_wps_pbc, 1, 0),
+	SHELL_CMD_ARG(wps_pin, NULL,
+		"Set and get wps pin.\n"
+		"[pin] No need input pin for get operation, for set operation is pin value.\n",
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_shell.c:2676]:
```suggestion
		"[pin] Only applicable for set.\n",
```

File: subsys/net/l2/wifi/wifi_shell.c [hunk 4]
```diff
@@ -2625,6 +2668,13 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
 		cmd_wifi_btm_query,
 		2, 0),
 #endif
+	SHELL_CMD_ARG(wps_pbc, NULL,
+		"Start wps pbc connection.\n",
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_shell.c:2672]:
```suggestion
		"Start a WPS PBC connection.\n",
```

File: subsys/net/l2/wifi/wifi_shell.c [hunk 5]
```diff
@@ -2625,6 +2668,13 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
...
+		cmd_wifi_wps_pbc, 1, 0),
+	SHELL_CMD_ARG(wps_pin, NULL,
+		"Set and get wps pin.\n"
```
Comment (krish2718) [inline subsys/net/l2/wifi/wifi_shell.c:2675]:
```suggestion
		"Set and get WPS pin.\n"
```

File: include/zephyr/net/wifi_mgmt.h [hunk 6]
```diff
@@ -1168,6 +1200,21 @@ struct wifi_mgmt_ops {
...
+	int (*wps_pbc)(const struct device *dev);
+	/** Set or get wps pin
+	 *
+	 * @param dev Pointer to the device structure for the driver instance
+	 * @param params wps pin operarion and pin number
+	 *
+	 * @return 0 if ok, < 0 if error
+	 */
+	int (*wps_pin)(const struct device *dev, struct wifi_wps_pin_params *params);
 };
```
Comment (krish2718) [inline include/zephyr/net/wifi_mgmt.h:1218]: do we really need two different APIs? can't we merge them as `wps_config` and extend `params` ?

File: include/zephyr/net/wifi_mgmt.h [hunk 7]
```diff
@@ -962,6 +974,26 @@ struct wifi_dpp_params {
...
+};
+
+/** Wi-Fi wps pin setup */
+struct wifi_wps_pin_params {
+	/** wps pin operation */
+	enum wifi_wps_pin_op oper;
+	/** get pin value*/
+	char get_pin[WIFI_WPS_PIN_MAX_LEN + 1];
+	/** set pin value*/
+	char set_pin[WIFI_WPS_PIN_MAX_LEN + 1];
```
Comment (krish2718) [inline include/zephyr/net/wifi_mgmt.h:994]: Can't we merge them as one `pin[WIFI_WPS_PIN_MAX_LEN + 1]`?  depending on the operation `pin` can either be input or output?

File: include/zephyr/net/wifi_mgmt.h [hunk 8]
```diff
@@ -962,6 +974,26 @@ struct wifi_dpp_params {
 	};
 };
 
+#define WIFI_WPS_PIN_MAX_LEN 8
+
+/** get/set operation for wps pin*/
```
Comment (krish2718) [inline include/zephyr/net/wifi_mgmt.h:979]:
```suggestion
/** Get/Set operation for WPS pin*/
```

File: modules/hostap/CMakeLists.txt [hunk 9]
```diff
@@ -285,7 +285,6 @@ zephyr_library_sources_ifdef(CONFIG_WIFI_NM_WPA_SUPPLICANT_WPS
 	${HOSTAP_SRC_BASE}/wps/wps_enrollee.c
 	${HOSTAP_SRC_BASE}/wps/wps_registrar.c
 	${HOSTAP_SRC_BASE}/crypto/dh_groups.c
-	${HOSTAP_SRC_BASE}/crypto/dh_group5.c
```
Comment (jukkar) [inline modules/hostap/CMakeLists.txt:288]: How is this change related to WPS?

File: subsys/net/l2/wifi/wifi_shell.c (hunk 1 above)
Comment (Rex-Chen-NXP) [reply subsys/net/l2/wifi/wifi_shell.c]: Updated as comment

File: subsys/net/l2/wifi/wifi_shell.c (hunk 2 above)
Comment (Rex-Chen-NXP) [reply subsys/net/l2/wifi/wifi_shell.c]: Updated as comment

File: subsys/net/l2/wifi/wifi_shell.c (hunk 4 above)
Comment (Rex-Chen-NXP) [reply subsys/net/l2/wifi/wifi_shell.c]: Updated as comment

File: subsys/net/l2/wifi/wifi_shell.c (hunk 5 above)
Comment (Rex-Chen-NXP) [reply subsys/net/l2/wifi/wifi_shell.c]: Updated as comment

File: include/zephyr/net/wifi_mgmt.h (hunk 7 above)
Comment (Rex-Chen-NXP) [reply include/zephyr/net/wifi_mgmt.h]: Changed as comment

File: include/zephyr/net/wifi_mgmt.h (hunk 8 above)
Comment (Rex-Chen-NXP) [reply include/zephyr/net/wifi_mgmt.h:973]: Updated as comment

File: modules/hostap/CMakeLists.txt (hunk 9 above)
Comment (Rex-Chen-NXP) [reply modules/hostap/CMakeLists.txt:288]: The reason why delete dh_group5.c is multiple definition compile errors when enable CONFIG_WIFI_NM_WPA_SUPPLICANT_WPS.
Actually, the I don't know the root reason why need use dh_groups.c but not dh_group5.c, the below 2 points support changed:
A. WPA3 also use  dh_groups.c [CONFIG_WIFI_NM_WPA_SUPPLICANT_WPA3].
B. I think dh_group5.c and dh_groups.c both are APIs support for ECDH,  dh_groups.c works ok for WPA3 SAE, I guess it should also will works ok for WPS.

**Comment by Rex-Chen-NXP** [comment]
Hi @jukkar  @krish2718 
Updated as your comments and test pass, pls help review again, and then I will resolve conflict files.
Thanks.

File: modules/hostap/CMakeLists.txt (hunk 9 above)
Comment (krish2718) [reply modules/hostap/CMakeLists.txt:288]: Independent of the reasoning, this is an independent change, so, has to be in a separate commit with commit log explaining the reason.

Comment (Rex-Chen-NXP) [reply modules/hostap/CMakeLists.txt:288]: Okay, if no build errors reported by CI of this pr after remove this change, will raise separate commit for it

File: subsys/net/l2/wifi/wifi_mgmt.c [hunk 10]
```diff
@@ -813,6 +813,21 @@ static int wifi_get_connection_params(uint32_t mgmt_request, struct net_if *ifac
 
 NET_MGMT_REGISTER_REQUEST_HANDLER(NET_REQUEST_WIFI_CONN_PARAMS, wifi_get_connection_params);
 
+static int wifi_wps_config(uint32_t mgmt_request, struct net_if *iface, void *data, size_t len)
+{
+	const struct device *dev = net_if_get_device(iface);
+	const struct wifi_mgmt_ops *const wifi_mgmt_api = get_wifi_api(iface);
+	struct wifi_wps_pin_params *params = data;
+
```
Comment (jukkar) [inline subsys/net/l2/wifi/wifi_mgmt.c:821]: We could also have a check that verifies that len is sizeof(*params)

Comment (Rex-Chen-NXP) [reply subsys/net/l2/wifi/wifi_mgmt.c:821]: I think we shouldn't check len is equal to sizeof(*params), because for WPS PBC case, len is 0

**Comment by Rex-Chen-NXP** [comment]
Hi @jukkar  @krish2718 
Could help review and approve PR?
Thanks.

File: include/zephyr/net/wifi_mgmt.h [hunk 11]
```diff
@@ -1010,6 +1016,24 @@ struct wifi_dpp_params {
...
+	WIFI_WPS_PIN_SET = 1,
+};
+
+/** Wi-Fi wps pin setup */
+struct wifi_wps_pin_params {
+	/** wps pin operation */
+	enum wifi_wps_pin_op oper;
+	/** pin value*/
+	char pin[WIFI_WPS_PIN_MAX_LEN + 1];
+};
```
Comment (krish2718) [inline include/zephyr/net/wifi_mgmt.h:1035]: An alternative is

```
struct wifi_wps_config_params {
  enum wifi_wps_oper oper;
  union {
    enum wifi_wps_pin_op oper;
    char pin[WIFI_WPS_PIN_MAX_LEN + 1];
  } pin_params;
};
```

We can even move the `pin_cfg` above to support, `WPS_PBC, WPS_PIN_SET, WPS_PIN_GET`.

Comment (Rex-Chen-NXP) [reply include/zephyr/net/wifi_mgmt.h]: Updated as comment

File: modules/hostap/src/supp_api.c [hunk 12]
```diff
@@ -1127,6 +1129,87 @@ int supplicant_get_wifi_conn_params(const struct device *dev,
...
+	return ret;
+}
+
+int supplicant_wps_config(const struct device *dev, struct wifi_wps_pin_params *params)
+{
+	if (params == NULL) {
+		return supplicant_wps_pbc(dev);
+	} else {
+		return supplicant_wps_pin(dev, params);
+	}
```
[the remark this answers is not in the document]
Comment (Rex-Chen-NXP) [reply modules/hostap/src/supp_api.c]: Updated as comment

**Comment by Rex-Chen-NXP** [comment]
Hi @jukkar  @krish2718 
Could you help to review again. thanks.

**Comment by MaochenWang1** [comment]
Hi @jukkar @krish2718  please review, thanks

**Comment by Rex-Chen-NXP** [comment]
Hi @jukkar @krish2718 
This PR block long time, pls help to merge if no other comment. Thanks!

**Review by jukkar — DISMISSED** [review]

**Review by jukkar — COMMENTED** [review]
BTW, if you send a new version, it would be nice to know what did you change to avoid reviewing the whole PR again.

**Comment by Rex-Chen-NXP** [comment]
> BTW, if you send a new version, it would be nice to know what did you change to avoid reviewing the whole PR again.

Yes, it should be.
For this update, only change log info to make it more reasonable.
subsys/net/l2/wifi/wifi_shell.c line 1773

**Review by jukkar — APPROVED** [review]

**Comment by Rex-Chen-NXP** [comment]
Hi @kartben @nashif @rlubos 
Could you help review, approve and merge the PR, as blocked long time.
Thanks!

**Comment by Rex-Chen-NXP** [comment]
Hi @jukkar @krish2718 
Could you help to approve and merge PR?

**Review by jukkar — APPROVED** [review]

**Review by dleach02 — APPROVED** [review]
